packages/contracts/contracts/SmartInvoiceFactoryV2.sol

```diff
@@ -0,0 +1,312 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/proxy/Clones.sol";
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "./interfaces/ISmartInvoiceFactoryV2.sol";
+import "./interfaces/ISmartInvoiceV2.sol";
+import "hardhat/console.sol";
+
+contract SmartInvoiceFactoryV2 is ISmartInvoiceFactoryV2, AccessControl {
+    uint256 public invoiceCount = 0;
+    mapping(uint256 => address) internal _invoices;
+
+    mapping(address => uint256) public resolutionRates;
+
+    bytes32 public constant ADMIN = keccak256("ADMIN");
+
+    /** @dev marks a deployed contract as a suitable implementation for additional escrow invoices formats */
+    /** @dev mapping(implementationType => mapping(implementationVersion => address)) */
+    mapping(bytes32 => mapping(uint256 => address)) public implementations;
+    mapping(bytes32 => uint256) public implementationsVersions;
```

Comment:
Should change implementationVersions to currentVersions or something like that for clarity